enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,247 @@
+The risk is small initially, when we have no such edges.
+If targeted blocking becomes widespread, a simple but tedious mitigation would be asking users to maintain their own graph-data (e.g. removing all targeted edge-blocking references) and build their own graph-data image.
+
+A more powerful but complicated fix would be documenting a process for running a local Telemetry aggregator and configuring the local update service to run queries against that aggregator.
+This documentation would need to cover mechanisms for adding any expected metrics that Red Hat currently adds in internal Telemetry which come up in targeted edge queries, such as [`subscription_labels`](#openshift-dedicated).
+
+Users who did not want to maintain their own graph-data or a local Telemetry aggregator could also deploy their local update service with the targeted edge blocking plugin disabled.
+This will leave all conditional edges *enabled*, with the local administrators taking responsibility for otherwise blocking or avoiding conditional edges to which their client cluster set may be vulnerable.
+
+#### Clusters not reporting data
+
+Clusters which have [opted out of uploading Telemetry][uploaded-telemetry-opt-out] or which run on restricted networks and so are unable to report Telemetry will default to the blocked state (more discussion on unknown ID handling in [the *query coverage* section](#query-coverage)).
+Being unable to report Telemetry while still being able to connect to Red Hat's hosted update service seems unlikely; if a cluster can do one (e.g. via a proxy), it should be able to do both.
+Clusters which connect to neither Telemetry nor Red Hat's hosted update service may run a local update service, and that is discussed in [the *divergent Prometheus services* section](#divergent-prometheus-services).
+
+Clusters which opt out of Telemetry are likely still able to connect to the update service.
+When they do, they will be [blocked as cache-misses](#enhanced-graph-data-schema-for-blocking-edges).
+I think that conservative approach is acceptable, because we cannot tell if the non-reporting cluster is vulnerable to the known issues or not.
+Local administrators should instead test their updates locally and make their own decisions about update safety, as discussed in [the *exposing blocking reasons* section](#exposing-blocking-reasons).
+
+#### Exposing blocking reasons
+
+This enhancement provides no mechanism for telling clients if or why a recommended update edge has been blocked, because [the Cincinnati graph format][cincinnati-spec] provides no mechanism for sharing metadata about recommended edges, or even the existence of not-recommended edges.
+Clients might be tempted to want that information to second-guess a graph-data decision to block the edge, but I am [not convinced the information is actionable](overriding-blocked-edges/README.md).
```

Comment:
OTA team needs a way to find out that information (not exposing it to the customers though). Currently in order to check if a cluster with particular ID has blocked updates we'd need to run all queries manually and check for cluster ID.

I see a few ways to resolve that:
* expose a metric for every blocking query so that OTA team could quickly check if ID is affected by targeted blocking. This might increase cardinality, so it needs to be carefully designed. 
* serializing cache in some temporary file so that we had a way to view it.

Reply:
> OTA team needs a way to find out that information...

Why do we need this?  I expect we will have cases where a user says "my cluster $ID does not see $EDGE".  But we can just make a Cincinnati request setting `id=$ID` to confirm that.  And if we are surprised by the results, we can run any related queries.  Shouldn't be too many possibilities for the given edge.  Does that seem unworkable to you?  Do you have other use cases?

Reply:
>I expect we will have cases where a user says "my cluster $ID does not see $EDGE". But we can just make a Cincinnati request setting id=$ID to confirm that.

That won't tell us which rule did it trigger. Running quieries may be tedious. In addition we might want to track the amount of clusters blocked by a particular rule (so that these won't be too aggressive)

Reply:
> That won't tell us which rule did it trigger.

Do we care about which rule triggered the block?  Also, the proposed schema does not allow for multiple `promql` expressions covering a single edge; it will be a single, possibly compound, expression to check.

> In addition we might want to track the amount of clusters blocked by a particular rule (so that these won't be too aggressive)

As described in the *Stranding supported clusters* section, I don't think we need to dip down into this level of detail to protect against stranding clusters.  The only other level I can see that makes sense is "does $SOME_PROMQL return the ID sets we expect?", which we can audit, when we see fit, by running the query directly without needing covering metrics or similar.

Reply:
I'm worried that some rules might mistakenly block more cluster than desired. OTOH we can add verbose logging to OSUS to trap this

Reply:
What's the workflow for auditing?  How do we determine expected match counts?  I'm not excited about committing to tooling to  support an undesigned process, bit obviously update service implementations can add logging, metrics, etc. without needing them blessed here